File: Day 06/Solution.py
# 209. Minimum Size Subarray Sum

# Given an array of positive integers nums and a positive integer target, return the minimal length of a subarray whose sum is greater than or equal to target. If there is no such subarray, return 0 instead.

from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        l = 0
        sum_ = 0
        res = float("inf")
        for r in range(len(nums)):
            sum_ += nums[r]
            while sum_ >= target:
                res = min(res, r - l + 1)
                sum_ -= nums[l]
                l += 1
        return 0 if res == float("inf") else res

# Example usage
if __name__ == "__main__":
    sol = Solution()
    print(sol.minSubArrayLen(7, [2,3,1,2,4,3]))  # Output: 2
    print(sol.minSubArrayLen(4, [1,4,4]))        # Output: 1
    print(sol.minSubArrayLen(11, [1,1,1,1,1,1,1,1]))  # Output: 0

File: Day 06/Solution.java
// 209. Minimum Size Subarray Sum

// Given an array of positive integers nums and a positive integer target, return the minimal length of a subarray whose sum is greater than or equal to target. If there is no such subarray, return 0 instead.

public class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int res = Integer.MAX_VALUE;
        int l = 0;
        int sum = 0;

        for (int r = 0; r < nums.length; r++) {
            sum += nums[r];
            while (sum >= target) {
                res = Math.min(res, r - l + 1);
                sum -= nums[l];
                l++;
            }
        }
        return res == Integer.MAX_VALUE ? 0 : res;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] arr1 = {2,3,1,2,4,3};
        int target1 = 7;
        System.out.println(sol.minSubArrayLen(target1, arr1)); // Output: 2

        int[] arr2 = {1,4,4};
        int target2 = 4;
        System.out.println(sol.minSubArrayLen(target2, arr2)); // Output: 1

        int[] arr3 = {1,1,1,1,1,1,1,1};
        int target3 = 11;
        System.out.println(sol.minSubArrayLen(target3, arr3)); // Output: 0
    }
}
